File: solve/1000+/1230.cc
// My Solution : https://github.com/ywcheong/solved-baekjoon

// Essential Headers
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>
#include <limits>

using namespace std;

typedef long long bigint;
const char eol = '\n';
const int INF = 987654321;

int solve(const string& original, const string& target) {
    int olen = original.length(), tlen = target.length();
    vector<vector<int>> memo(olen + 1, vector<int>(tlen + 1, -1));

    for (int o_sublen = 0; o_sublen <= olen; ++o_sublen) {
        for (int t_sublen = 0; t_sublen <= tlen; ++t_sublen) {
            int result;

            if (o_sublen == 0) {
                result = (t_sublen == 0) ? 0 : 1;
            } else {
                char o_char = original[o_sublen - 1];
                result = INF;

                for (int t_index = 0; t_index < t_sublen; ++t_index) {
                    char t_char = target[t_index];

                    if (o_char == t_char) {
                        int this_result = memo[o_sublen - 1][t_index];

                        if (t_index < t_sublen - 1) {
                            ++this_result;
                        }

                        result = min(result, this_result);
                    }
                }
            }

            memo[o_sublen][t_sublen] = result;
        }
    }

    int output = memo[olen][tlen];
    return (output == INF) ? -1 : output;
}

void test() {
    auto check = [](int left, int right) {
        if (left == right) {
            cout << "Test pass" << eol;
        } else {
            cout << "Test fail: " << left << " != " << right << eol;
        }
    };

    check(solve("hello fine", "hello, how are you? I'm fine thank you and you?"), 2);
    check(solve("aaaaa", "ababababa"), 4);
    check(solve("no way", "No way!"), -1);
    check(solve("abcefijklmnopuvwxz", "abcdefghijklmnopqrstuvwxyz"), 4);
    check(solve("B", "ABC"), 2);
    check(solve("A", "ABC"), 1);
    check(solve("C", "ABC"), 1);
    check(solve("C", "ABCDE"), 2);
    check(solve("", ""), 0);
    check(solve("A", ""), -1);
    check(solve("", "A"), 1);
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    string original, target;
    getline(cin, original);
    getline(cin, target);
    cout << solve(original, target) << eol;

    // Uncomment the following line to run tests
    // test();

    return 0;
}

File: solve/1000+/1230.py
# My Solution : https://github.com/ywcheong/solved-baekjoon

import sys


def input_one(given_type):
    return given_type(sys.stdin.readline()[:-1])


def input_list(given_type):
    return list(map(given_type, input_one(str).split()))


def solve(original, target):
    """write your logic here"""
    olen, tlen = len(original), len(target)
    memo = [[None] * (tlen + 1) for _ in range(olen + 1)]

    for o_sublen in range(olen + 1):
        for t_sublen in range(tlen + 1):
            result = None

            if o_sublen == 0:
                if t_sublen == 0:
                    result = 0
                else:
                    result = 1
            else:
                o_char = original[o_sublen - 1]
                result = float("inf")

                for t_index in range(t_sublen):
                    t_char = target[t_index]

                    if o_char == t_char:
                        this_result = memo[o_sublen - 1][t_index]

                        if t_index < t_sublen - 1:
                            this_result += 1

                        result = min(result, this_result)

            memo[o_sublen][t_sublen] = result

    output = memo[olen][tlen]
    if output == float("inf"):
        return -1
    return output


def test():
    def check(left, right):
        if left == right:
            print(f"Test pass")
        else:
            print(f"Test fail: {left} != {right}")

    """write your test here"""
    check(solve("hello fine", "hello, how are you? I'm fine thank you and you?"), 2)
    check(solve("aaaaa", "ababababa"), 4)
    check(solve("no way", "No way!"), -1)
    check(solve("abcefijklmnopuvwxz", "abcdefghijklmnopqrstuvwxyz"), 4)
    check(solve("B", "ABC"), 2)
    check(solve("A", "ABC"), 1)
    check(solve("C", "ABC"), 1)
    check(solve("C", "ABCDE"), 2)
    check(solve("", ""), 0)
    check(solve("A", ""), -1)
    check(solve("", "A"), 1)


def main():
    """write your i/o here"""
    original = input_one(str)
    target = input_one(str)
    print(solve(original, target))


# test()
main()
